# Weekly assignments week 4-5

Weekly assignment 4: machine learning (1/2)

Introduction: describe which dataset and variables you selected and why. 

I have decided to go with the employee attrition dataset. I'm curious to know what potentially could cause employee attrition. I have picked seven variables that, in my opinion, are best to predict employee attrition. I have choosen the following variables: PerformanceRating, Education, EnvironmentSatisfaction, JobInvolvement, JobSatisfaction, PerformanceRating and WorkLifeBalance. 

In [2]:
import seaborn as sns
import numpy as np
import sklearn as sk
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split

In [3]:
df = pd.read_csv('employee.csv') 
df = df.dropna()
df.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1            80                 0   
1  ...                         4            80                 1   
2  ...                         2            80                 0   
3  ...                         3            80                 0   
4  ...                         4            80                 1   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                      0               1               6   
1                 10                      3               3              10   
2                  7                      3               3               0   
3                  8                      3               3               8   
4                  6                      3               3               2   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                     5  
1                  7                        1                     7  
2                  0                        0                     0  
3                  7                        3                     0  
4                  2                        2                     2  

[5 rows x 35 columns]

In [4]:
for col in df.columns:
    print(col)

Age
Attrition
BusinessTravel
DailyRate
Department
DistanceFromHome
Education
EducationField
EmployeeCount
EmployeeNumber
EnvironmentSatisfaction
Gender
HourlyRate
JobInvolvement
JobLevel
JobRole
JobSatisfaction
MaritalStatus
MonthlyIncome
MonthlyRate
NumCompaniesWorked
Over18
OverTime
PercentSalaryHike
PerformanceRating
RelationshipSatisfaction
StandardHours
StockOptionLevel
TotalWorkingYears
TrainingTimesLastYear
WorkLifeBalance
YearsAtCompany
YearsInCurrentRole
YearsSinceLastPromotion
YearsWithCurrManager


In an earlier stage I have used the dropna() code. It removes the rows that contains NULL values. Furthermore I have picked seven variables applicable to the prediction. 

In [5]:
df_subset = df[['Attrition', 'PerformanceRating', 'Education', 'EnvironmentSatisfaction', 'JobInvolvement', 'JobSatisfaction', 'PerformanceRating', 'WorkLifeBalance']]
df_subset.head()

Attrition  PerformanceRating  Education  EnvironmentSatisfaction  \
0       Yes                  3          2                        2   
1        No                  4          1                        3   
2       Yes                  3          2                        4   
3        No                  3          4                        4   
4        No                  3          1                        1   

   JobInvolvement  JobSatisfaction  PerformanceRating  WorkLifeBalance  
0               3                4                  3                1  
1               2                2                  4                3  
2               2                3                  3                3  
3               3                3                  3                3  
4               3                2                  3                3

In [6]:
df_subset['Attrition'].value_counts()

No     1233
Yes     237
Name: Attrition, dtype: int64

Weekly assignment 5: machine learning (2/2)

In [7]:
from sklearn.preprocessing import normalize

X = df_subset[['PerformanceRating', 'Education', 'EnvironmentSatisfaction', 'JobInvolvement', 'JobSatisfaction', 'PerformanceRating', 'WorkLifeBalance']] 
X = normalize(X) 
y = df_subset['Attrition'] 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1) 

Explanation algorithm:

In [8]:
from sklearn.neighbors import KNeighborsClassifier

knn = KNeighborsClassifier(n_neighbors=5) 
knn = knn.fit(X_train, y_train) 
knn.score(X_test, y_test) 

0.8095238095238095

81% of attrition is predicted accurately.

In [9]:
from sklearn.metrics import confusion_matrix

y_test_pred = knn.predict(X_test) 
cm = confusion_matrix(y_test, y_test_pred) 
cm

array([[352,  12],
       [ 72,   5]], dtype=int64)

In [10]:
conf_matrix = pd.DataFrame(cm, index=['False_actual','True_actual'], columns = ['False_predicted','True_predicted']) 
conf_matrix

False_predicted  True_predicted
False_actual              352              12
True_actual                72               5

Interpret results: 

The way to read this is that of the false_actual, 352 are correctly predicted as 'false', 12 are instead predicted as 'true'. The recall and precision for the category false_actual is:

- Recall = 352 / 352 + 12 = .97
- Precision = 352 / 352 + 72 = .83